texts/0004-wry-raw-window-handle.md

```diff
@@ -0,0 +1,113 @@
+
+# Motivation
+
+This has been requested in [tauri-apps/wry#677](https://github.com/tauri-apps/wry/issues/677) a long time ago and at work at CrabNebula,
+we have been contracted to research and develop an integration of wry and wgpu/bevy. We had a successful results and it aligned with [tauri-apps/wry#677](https://github.com/tauri-apps/wry/issues/677)
+so we did an initial implemention of this RFC for Windows which can be found here https://github.com/crabnebula-dev/wry/commits/feat/embedded-webview and we think
+it is a good oprtunity to upstream work.
+
+# Guide-level explanation
+There is two groups of window handles we will need to take in order to support the current platforms we support:
+
+#### Raw window handle:
+This group could leverage the `raw-window-handle` crate; this includes macOS, Windows and Linux (x11) platforms.
+  - Using winit, tao, egui or any other windowing library that
+    implements [`raw_window_handle::HasWindowHandle`](https://docs.rs/raw-window-handle/latest/raw_window_handle/trait.HasWindowHandle.html) trait
+    (or previous versions of `raw-window-handle` crate):
+    ```rs
+    let window = Window::new(&evl);
+    let raw_handle = window.raw_window_handle();
+    let webview = WebView::new(raw_handle);  
+    ```
+    
+  - Using raw win32:
+    ```rs
+    let hwnd = CreateWindowExW(/* args */);
+    let raw_handle =  RawWindowHandle::Win32(Win32WindowHandle {
+      hwnd,
+      hinstance: None,
+    });
+    let webview = WebView::new(raw_handle);  
+    ```
+    
+  - Using raw AppKit NSWindow:
+    ```rs
+    let ns_window = msg_send![/* args */];
+    let raw_handle =  RawWindowHandle::Win32(AppKitWindowHandle  {
+      ns_view: ns_window.contentView(),
+    });
+    let webview = WebView::new(raw_handle);  
+    ```
+  - Using raw X11:
+    ```rs
+    let window = XCreateWindow(/* args */);
+    let raw_handle =  RawWindowHandle::Xlib(XlibWindowHandle::new(window);
+    let webview = WebView::new(raw_handle);  
+    ```
+
+#### Gtk handle:
+This group leverages the gtk types, includes Linux platform only (X11 and Wayland). 
+  - Using raw gtk:
+    ```rs
+    let gtk_window = gtk::ApplicationWindow::new(/* args */);
+    let webview = WebView::new_gtk(&gtk_window);  
+    ```
+  - Using tao:
+    ```rs
+    let window = Window::new(/* args */);
+    let gtk_window = window.gtk_window();
+    let webview = WebView::new_gtk(&gtk_window);  
+    ```
+
+
+# Reference-level explanation
+
+There is 2 APIs this RFC is proposing:
+- `Webview::new(parent_raw_window_handle)` initializes the Webview as a child in the passed window handle, this allows for multiple webviews in
+  the same window handle passed. 
+  - Supported platforms: Windows, macOS and Linux (X11)
+  - Supports `webview.set_position()/webview.set_size()` to change the position and size of the webview child inside the parent later on.
+- `Webview::new_in(raw_window_handle)/Webview::new_in_gtk(gtk_handle)` (current behavior)
```

Comment:
I don't like names of `Webview::new` and `Webview::new_in`. It will be confusing if the variables aren't explicit enough. We should think about better naming for this.

Reply:
I agree, how about `Webview::new(window)` and `Webview::new_as_child_in(parent)`

Reply:
I think we should keep `Webview::new()` as what we have right now which initializes the webview directly in the passed window handle.
And we can add an attribute like `child_view: bool` for it to be child window/view in the parent window.

As of `new_in_gtk`, I think this will also bring bad DX on Linux since users will have to call a different function.
I'm thinking about an enum like `WebviewWindow` to include both raw handle type and gtk type. So the signature of the new method would be `fn new(window: Into<WebviewWindow>)`. Users could pass their existing type without worry what have been changed.

Reply:
Okay let's stick to `Webview::new`, `new_gtk`, `new_as_child` then.